### PR TITLE
Feat: add attention: 1. score function: additive, location_sensitive, scaled_dot_product; 2. self-attention: MultiHeadAttention

### PR DESCRIPTION
**Pull request by weedge**
# attention
image

image

### Additive attention
image

- 在预测词元时，如果不是所有输入词元都是相关的，那么具有Bahdanau注意力的循环神经网络编码器-解码器会有选择地统计输入序列的不同部分。这是通过将上下文变量视为加性注意力池化的输出来实现的。
- 在循环神经网络编码器-解码器中，Bahdanau注意力将上一时间步的解码器隐状态视为查询，在所有时间步的编码器隐状态同时视为键和值。

原始论文中应用在NMT任务seq2seq模型的decoder中

[Neural Machine Translation by Jointly Learning to Align and Translate](https://arxiv.org/abs/1409.0473v7)(RNN(LSTM)+Additive Attention(Bahdanau Attention) in NMT)

image

image

image

image

### Multiplicative(Dot-Product) attention

原始论文中应用在NMT任务seq2seq模型的decoder中
[Effective Approaches to Attention-based Neural Machine Translation](https://arxiv.org/abs/1508.04025)(RNN(LSTM)+(global/local)multiplicative attention(Luong Attention))

将attention model分成global 和 local attention, 两者计算差不多，local attention 通过滑动窗口计算
image
image

image

### Scaled Dot-Product Attention

❤[**Attention Is All You Need**](https://arxiv.org/abs/1706.03762)❤ (Transformer,based solely on attention mechanisms, dispensing with recurrence and convolutions entirely. **self attention(Scaled Dot-Product Attention(SDPA), Multi-Head Attention)**)

原始论文中应用在NMT任务seq2seq模型的encoder-decoder中
image
image

### Multi-Head Attention
image

image

## self-attention
在深度学习中，经常使用卷积神经网络（CNN）或循环神经网络（RNN）对序列进行编码。 想象一下，有了注意力机制之后，我们将词元序列输入注意力池化中， 以便同一组词元同时充当查询、键和值。 具体来说，每个查询都会关注所有的键－值对并生成一个注意力输出。 由于查询、键和值来自同一组输入，因此被称为 自注意力（self-attention）  也被称为内部注意力（intra-attention）

## positional-encoding

在处理token序列时，循环神经网络是逐个的重复地处理token， 而自注意力则因为并行计算而放弃了顺序操作。 为了使用序列的顺序信息，通过在输入表示中添加 位置编码（positional encoding）来注入绝对的或相对的位置信息。 位置编码可以通过学习得到也可以直接固定得到。 

## seq2seq with attention (rnn+attention, cnn+attention, self-attention(all))
image

### rnn vs cnn vs  self-attention
- 在自注意力中，查询、键和值都来自同一组输入。
- 卷积神经网络和自注意力都拥有并行计算的优势，而且自注意力的最大路径长度最短。但是因为其计算复杂度是关于序列长度的二次方，所以在很长的序列中计算会非常慢。
- 在处理token序列时，循环神经网络是逐个的重复地处理token， 而自注意力则因为并行计算而放弃了顺序操作。 自注意力为了使用序列的顺序信息，可以通过在输入表示中添加位置编码，来注入绝对的或相对的位置信息。

image
image

## summary
image

# Transformer

❤[**Attention Is All You Need**](https://arxiv.org/abs/1706.03762)❤ (Transformer,based solely on attention mechanisms, dispensing with recurrence and convolutions entirely. **self attention(Scaled Dot-Product Attention(SDPA), Multi-Head Attention)**)

## self-attention
Transformer 以三种不同的方式使用多头注意力：

- Encoder-Decoder attention: 在“编码器-解码器注意力”层中，查询来自前一个解码器层，内存键和值来自编码器的输出。这允许解码器中的每个位置都参与输入序列中的所有位置。这模仿了序列到序列模型中典型的编码器-解码器注意机制，例如 [Google's Neural Machine Translation System: Bridging the Gap between Human and Machine Translation](https://arxiv.org/abs/1609.08144).

image

- Encoder self-attention: 编码器包含自注意力层。在自注意力层中，所有键、值和查询都来自同一位置，在本例中是编码器中前一层的输出。编码器中的每个位置可以关注编码器上一层中的所有位置.

- MaskedDecoder self-attention: 类似地，解码器中的自注意力层允许解码器中的每个位置关注解码器中直到并包括该位置的所有位置。我们需要防止解码器中的左向信息流以保留自回归属性。我们通过屏蔽（设置为 -∞) softmax 输入中对应于非法连接的所有值。

image

## positional-encoding
原始论文中使用基于正弦函数和余弦函数的固定位置编码

---

更多细节见：https://github.com/weedge/doraemon-nb/blob/main/AnnotatedTransformer.ipynb

### CODE REVIEW

**Comment by weedge** [comment]
next: Transformers

**Comment by weedge** [comment]
more:
## longer context length
the best ways for extending a pretrained model to longer context length:
- [ring attention](https://x.com/Haojun_Zhao14/status/1815419356408336738)
- [YaRN](https://arxiv.org/abs/2309.00071) 
- [rope scaling](https://arxiv.org/abs/2309.16039)
---
- [Infini-attention](https://arxiv.org/abs/2404.07143): https://weedge.github.io/post/paper/transformer/infini_attention/
- https://huggingface.co/blog/infini-attention

**Comment by weedge** [comment]
## Key-Value (KV) cache (减少计算，加快forward推理速度(训练和inference))

在 Transformer 解码器中，由于 token 的注意力依赖于前面的 token，因此不会重新计算前面的上下文，而是缓存其 Key 和 Value。这可以显着加快推理速度，但随着序列长度和模型维度的增长，可能会带来昂贵的内存开销。

原始Transformer使用 Multi-Head Attention （MHA） 多头注意力 , 在这种背景下，为了该井来改进MHA  K/V cache， 提出了不同的改进self-attention结构：

### Multi-Query Attention  多查询注意力 (MQA)
- [2019. Fast Transformer Decoding: One Write-Head is All You Need](https://arxiv.org/abs/1911.02150)

### Grouped-Query Attention  分组查询注意力 (GQA)
- [2023. Gqa: Training generalized multi-query transformer models from multi-head checkpoints](https://arxiv.org/abs/2305.13245)

### Multi-Head Latent Attention 多头潜在注意力 (MLA)
- [2024. DeepSeek-V2: A Strong, Economical, and Efficient Mixture-of-Experts Language Model](https://arxiv.org/abs/2405.04434)
